File: interfaces/python/tests/general/interrupt.py
"""
Big lasso problem
"""

import numpy as np
import scipy.sparse as spa
import osqp

# Problem dimensions and sparity
n = 100
m = 10*n
dens_lvl = 0.5

np.random.seed(1)
# Generate data
Ad = spa.random(m, n, density=dens_lvl)
x_true = np.multiply((np.random.rand(n) > 0.5).astype(float),
                     np.random.randn(n)) / np.sqrt(n)
bd = Ad.dot(x_true) + .5*np.random.randn(m)
gamma = 1.

#       minimize	y.T * y + gamma * np.ones(n).T * t
#       subject to  y = Ax
#                   -t <= x <= t
P = spa.block_diag((spa.csc_matrix((n, n)), 2*spa.eye(m),
                    spa.csc_matrix((n, n))), format='csc')
q = np.append(np.zeros(m + n), gamma*np.ones(n))
In = spa.eye(n)
Onm = spa.csc_matrix((n, m))
A = spa.vstack([spa.hstack([Ad, -spa.eye(m),
                            spa.csc_matrix((m, n))]),
                spa.hstack([In, Onm, -In]),
                spa.hstack([In, Onm, In])]).tocsc()
l = np.hstack([bd, -np.inf * np.ones(n), np.zeros(n)])
u = np.hstack([bd, np.zeros(n), np.inf * np.ones(n)])

osqp_opts = {'check_termination': 0,
             'polish': False,
             'scaling': True,
             'max_iter': 50000}

model = osqp.OSQP()
model.setup(P=P, q=q, A=A, l=l, u=u, **osqp_opts)
res_osqp = model.solve()

File: interfaces/matlab/examples/interrupt.m
%% Big lasso problem

% Problem dimensions and sparity
n = 100;
m = 10*n;
dens_lvl = 0.5;

rng(1);
% Generate data
Ad = sprandn(m, n, dens_lvl);
x_true = (rand(n, 1) > 0.5).*randn(n, 1) / sqrt(n);
bd = Ad*x_true + .5*randn(m, 1);
gamma = 1;

%   minimize	y.T * y + gamma * np.ones(n).T * t
%   subject to  y = Ax
%               -t <= x <= t
P = blkdiag(sparse(n, n), 2*speye(m), sparse(n, n));
q = [zeros(m+n,1); gamma*ones(n,1)];
In = speye(n);
Onm = sparse(n, m);
A = [Ad, -eye(m), sparse(m, n);
     In, Onm, -In;
     In, Onm, In];
l = [bd; -Inf*ones(n,1); zeros(n,1)];
u = [bd; zeros(n,1); Inf*ones(n,1)];

solver = osqp;
solver.setup(P, q, A, l, u, ...
             'check_termination', 0, 'polish', 0, 'max_iter', 50000);
res = solver.solve();

% PRESS CTRL-C BEFORE REACHING 50K ITERATIONS
